### PR TITLE
Switch golang.org/x/crypto to gravitational fork

### LINKED ISSUE #10918: Newer OpenSSH clients are dropping support for `ssh-rsa-cert-v01`
### Description

**What happened**:

Fedora 35 bundles OpenSSH 8.7 and the default system crypto policy has been changed to remove `ssh-rsa` and `ssh-rsa-cert-v01` from the approved list of `PubKeyAcceptedAlgorithms`:

```
[vagrant@fedora35 ~]$ grep PubkeyAcceptedAlgorithms /etc/crypto-policies/back-ends/openssh.config | tr "," "\n"
PubkeyAcceptedAlgorithms ecdsa-sha2-nistp256
ecdsa-sha2-nistp256-cert-v01@openssh.com
sk-ecdsa-sha2-nistp256@openssh.com
sk-ecdsa-sha2-nistp256-cert-v01@openssh.com
ecdsa-sha2-nistp384
ecdsa-sha2-nistp384-cert-v01@openssh.com
ecdsa-sha2-nistp521
ecdsa-sha2-nistp521-cert-v01@openssh.com
ssh-ed25519
ssh-ed25519-cert-v01@openssh.com
sk-ssh-ed25519@openssh.com
sk-ssh-ed25519-cert-v01@openssh.com
rsa-sha2-256
rsa-sha2-256-cert-v01@openssh.com
rsa-sha2-512
rsa-sha2-512-cert-v01@openssh.com
```

This breaks `ssh` connectivity to Teleport clusters when using `tsh config` with an auto-generated config file:

```
[vagrant@fedora35 ~]$ ssh ubuntu@ip-10-0-131-171
ubuntu@gus.teleportdemo.com: Permission denied (publickey).
ERROR: exit status 255

kex_exchange_identification: Connection closed by remote host
Connection closed by UNKNOWN port 65535
```

Expanding the `ProxyCommand` out to run `ssh` manually and using `-v`, we can see that the underlying error is actually `no mutual signature algorithm`:

```diff
[vagrant@fedora35 ~]$ ssh -v -p 3023 ubuntu@example.teleport.com -s proxy:ip-10-0-131-171:3022
OpenSSH_8.7p1, OpenSSL 1.1.1l  FIPS 24 Aug 2021
debug1: Reading configuration data /home/vagrant/.ssh/config
debug1: /home/vagrant/.ssh/config line 6: Applying options for example.teleport.com
debug1: /home/vagrant/.ssh/config line 13: Skipping Host block because of negated match for example.teleport.com
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: Reading configuration data /etc/ssh/ssh_config.d/50-redhat.conf
debug1: Reading configuration data /etc/crypto-policies/back-ends/openssh.config
debug1: configuration requests final Match pass
debug1: re-parsing configuration
debug1: Reading configuration data /home/vagrant/.ssh/config
debug1: /home/vagrant/.ssh/config line 6: Applying options for example.teleport.com
debug1: /home/vagrant/.ssh/config line 13: Skipping Host block because of negated match for example.teleport.com
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: Reading configuration data /etc/ssh/ssh_config.d/50-redhat.conf
debug1: Reading configuration data /etc/crypto-policies/back-ends/openssh.config
debug1: Connecting to example.teleport.com [1.2.3.4] port 3023.
debug1: Connection established.
debug1: identity file /home/vagrant/.tsh/keys/example.teleport.com/gus-local type 0
debug1: certificate file /home/vagrant/.tsh/keys/example.teleport.com/gus-local-ssh/purple-cert.pub type 4
debug1: Local version string SSH-2.0-OpenSSH_8.7
debug1: Remote protocol version 2.0, remote software version Teleport
debug1: compat_banner: no match: Teleport
debug1: Authenticating to example.teleport.com:3023 as 'ubuntu'
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: curve25519-sha256@libssh.org
debug1: kex: host key algorithm: rsa-sha2-512-cert-v01@openssh.com
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: kex: curve25519-sha256@libssh.org need=64 dh_need=64
debug1: kex: curve25519-sha256@libssh.org need=64 dh_need=64
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: SSH2_MSG_KEX_ECDH_REPLY received
debug1: Server host certificate: ssh-rsa-cert-v01@openssh.com SHA256:3Sl9kSR4U+ovsjM6OKaD4JPuAxFBcWlGelnAQOm6P0E, serial 0 ID "" CA ssh-rsa SHA256:s4nqYP8iLSHc9veGtJEie8ndQM6JDDxUvf6EuvCfFCo valid after 2022-01-28T17:59:23
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug1: checking without port identifier
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug1: Host 'example.teleport.com' is known and matches the RSA-CERT host certificate.
debug1: Found CA key in /home/vagrant/.tsh/known_hosts:1
debug1: found matching key w/out port
debug1: rekey out after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey in after 134217728 blocks
debug1: Will attempt key: /home/vagrant/.tsh/keys/example.teleport.com/gus-local-ssh/purple-cert.pub RSA-CERT SHA256:s8tjExcDL3bXMT9Wv9tKkO08tI8ZUZas0ZWd1rNpMmI explicit
debug1: Will attempt key: /home/vagrant/.tsh/keys/example.teleport.com/gus-local RSA SHA256:s8tjExcDL3bXMT9Wv9tKkO08tI8ZUZas0ZWd1rNpMmI explicit agent
debug1: Will attempt key: teleport:gus-local RSA-CERT SHA256:s8tjExcDL3bXMT9Wv9tKkO08tI8ZUZas0ZWd1rNpMmI agent
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey
debug1: Next authentication method: publickey
debug1: Offering public key: /home/vagrant/.tsh/keys/example.teleport.com/gus-local-ssh/purple-cert.pub RSA-CERT SHA256:s8tjExcDL3bXMT9Wv9tKkO08tI8ZUZas0ZWd1rNpMmI explicit
-debug1: send_pubkey_test: no mutual signature algorithm
debug1: Offering public key: /home/vagrant/.tsh/keys/example.teleport.com/gus-local RSA SHA256:s8tjExcDL3bXMT9Wv9tKkO08tI8ZUZas0ZWd1rNpMmI explicit agent
-debug1: send_pubkey_test: no mutual signature algorithm
debug1: Offering public key: teleport:gus-local RSA-CERT SHA256:s8tjExcDL3bXMT9Wv9tKkO08tI8ZUZas0ZWd1rNpMmI agent
-debug1: send_pubkey_test: no mutual signature algorithm
debug1: No more authentication methods to try.
ubuntu@example.teleport.com: Permission denied (publickey).
```

### Workaround
The current workaround is to add these two lines to the first section of the generated `~/.ssh/config` file:

```diff
#
# Begin generated Teleport configuration for example.teleport.com:443 from `tsh config`
#

# Common flags for all purple hosts
Host *.purple example.teleport.com
    UserKnownHostsFile "/home/vagrant/.tsh/known_hosts"
    IdentityFile "/home/vagrant/.tsh/keys/example.teleport.com/gus-local"
    CertificateFile "/home/vagrant/.tsh/keys/example.teleport.com/gus-local-ssh/purple-cert.pub"
+   HostKeyAlgorithms ssh-rsa-cert-v01@openssh.com
+   PubkeyAcceptedAlgorithms +ssh-rsa-cert-v01@openssh.com

# Flags for all purple hosts except the proxy
Host *.purple !example.teleport.com
    Port 3022
    ProxyCommand "/usr/local/bin/tsh" proxy ssh --cluster=purple --proxy=example.teleport.com %r@%h:%p

# End generated Teleport configuration
```

On Fedora, you can also run `sudo update-crypto-policies --set LEGACY` to switch back to the older crypto policies which still trust `ssh-rsa-cert-v01` and `ssh-rsa`.

**What you expected to happen**: Connectivity should work out of the box. Either Teleport should upgrade to a newer SSH certificate format, or automatically add the `PubkeyAcceptedKeyTypes` line to generated configs with `tsh config`.

### Reproduction Steps
Run above steps on a Fedora 35 machine.

### Server Details
- Teleport version (run `teleport version`): `Teleport v8.3.4 git:v8.3.4-0-g010bea1 go1.17.3`
- Server OS (e.g. from `/etc/os-release`): Fedora 35
- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware): Vagrant
- Additional details:

### Client Details
- Tsh version (`tsh version`): `Teleport v8.3.4 git:v8.3.4-0-g010bea1 go1.17.3`

### LINKED ISSUE #17046: Investigate and resolve Teleport and CentOS 7 integration issues
Investigate and resolve any issues that exist when using Teleport with OpenSSH on CentOS 7.

* Ensure Teleport SSH servers speak the cryptographic algorithms supported by the version of OpenSSH that ships with CentOS 7.
* Ensure that `tsh` generates configuration that can be used with `ssh` when connecting to Teleport servers.

### LINKED ISSUE #17197: OpenSSH server compatibility issues
Expected behavior:
OpenSSH servers can be accessed via teleport using tsh, the Teleport WebUI, or OpenSSH

Current behavior:

| client                            | sshd version      | working | server error |
| ----------------------|-----------------| --------|-------------|
| tsh 11.0.0-alpha.2        | OpenSSH_7.4p1 | ❌         | `unsupported public key algorithm: [rsa-sha2-256-cert-v01@openssh.com]` |
| tsh 10.0.0                      | OpenSSH_7.4p1 | ❌         | `unsupported public key algorithm: [rsa-sha2-256-cert-v01@openssh.com]` |
| Web UI 11.0.0-alpha.2 | OpenSSH_7.4p1 | ❌         | `unsupported public key algorithm: [rsa-sha2-256-cert-v01@openssh.com]` |
| Web UI 10.0.0              | OpenSSH_7.4p1 | ❌         | `unsupported public key algorithm: [rsa-sha2-256-cert-v01@openssh.com]` |
| ssh OpenSSH_8.6p1   | OpenSSH_7.4p1 | ✅         |
| tsh 11.0.0-alpha.2        | OpenSSH_8.9p1 | ✅         |
| Web UI 11.0.0-alpha.2 | OpenSSH_8.9p1 | ✅         |
| ssh OpenSSH_8.6p1   | OpenSSH_8.9p1 | ❌         | `no matching host key type found. Their offer: ssh-rsa-cert-v01@openssh.com` |

Same behaviour on local and leaf cluster.

### PR DESCRIPTION
**Pull request by jakule**
This PR switches Teleport from using `golang.org/x/crypto` to our internal fork `github.com/gravitational/crypto` to work around the issue with OpenSSH < 7.6.
Recently, Go crypto added support for OpenSSH 8.5+ https://github.com/golang/crypto/commit/6fad3dfc18918c2ac9c112e46b32473bd2e5e2f9, but they also broke the older OpenSSH compatibility. Our fork works with older and "modern" OpenSSH, including this https://github.com/gravitational/crypto/commit/903e65687e5969ad898fcae60f26bc897c7b2351 change (not yet available upstream).

Closes https://github.com/gravitational/teleport/issues/10918
Closes https://github.com/gravitational/teleport/issues/17197
Closes https://github.com/gravitational/teleport/issues/17046

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -371,4 +366,6 @@ replace (
 	github.com/pkg/sftp => github.com/gravitational/sftp v1.13.6-0.20220927202521-0e74d42f8055
 	github.com/sirupsen/logrus => github.com/gravitational/logrus v1.4.4-0.20210817004754-047e20245621
 	github.com/vulcand/predicate => github.com/gravitational/predicate v1.3.0
+	// Use our internal crypto fork, to work around the issue with OpenSSH <= 7.6 mentioned here: https://github.com/golang/go/issues/53391
```
Comment (hatched) [inline go.mod:369]: Do we have an issue somewhere to track reverting this change once this issue has been resolved upstream?

**Review by hatched — APPROVED** [review]

**Comment by jakule** [comment]
@fspmarshall Friendly ping.

**Review by zmb3 — APPROVED** [review]

**Comment by zmb3** [comment]
cc @rosstimothy for awareness - we'll have to keep our fork up to date as upstream changes

File: go.mod [hunk 2]
```diff
@@ -122,6 +122,7 @@ require (
 	go.opentelemetry.io/otel/sdk v1.11.2
 	go.opentelemetry.io/otel/trace v1.11.2
 	go.opentelemetry.io/proto/otlp v0.19.0
+	golang.org/x/crypto v0.2.0
```
Comment (codingllama) [inline go.mod:125]:
```suggestion
	golang.org/x/crypto v0.2.0 // replaced
```

so it's easier to spot when we do dependency updates. Same for api.

**Review by codingllama — CHANGES_REQUESTED** [review]
Could you update the following files as well?

* api/go.mod -> needs the same replace as base go.mod, and it has the same "do not update" warning (https://github.com/gravitational/teleport/blob/master/api/go.mod#L32-L36).

* .github/dependabot.yml -> please move crypto to the forked/replaced section (https://github.com/gravitational/teleport/blob/master/.github/dependabot.yml#L19), for both go.mod and api/go.mod entries.

**Comment by jakule** [comment]
> Does this change allow us to update dependencies which in turn depend on > `x/crypto v0.2.0`?

@rosstimothy Yes, with this change, we can use `x/crypto` HEAD.

**Comment by codingllama** [comment]
> Could you update the following files as well?
> 
> * api/go.mod -> needs the same replace as base go.mod, and it has the same "do not update" warning (https://github.com/gravitational/teleport/blob/master/api/go.mod#L32-L36).
> * .github/dependabot.yml -> please move crypto to the forked/replaced section (https://github.com/gravitational/teleport/blob/master/.github/dependabot.yml#L19), for both go.mod and api/go.mod entries.

Missing dependabot.yml?

**Comment by jakule** [comment]
@codingllama Sorry, I missed that. Fixed https://github.com/gravitational/teleport/pull/19579/commits/981e541dd172b497dcc483599ff53ed0afbb1f0b

**Review by codingllama — APPROVED** [review]

**Pull request merged**